### PR TITLE
[NVIDIA] [feat] Integrate flashinfer Trtllmgen bf16 moe

### PR DESCRIPTION
**Pull request by jiahanc**
## Purpose
- Integrate flashinfer trtllm-gen BF16 moe to supported models
- Fixed flashinfer trtllm-gen moe SM check
## Test Plan
`VLLM_USE_FLASHINFER_MOE_FP16=1 VLLM_FLASHINFER_MOE_BACKEND=latency vllm serve Qwen/Qwen3-Next-80B-A3B-Instruct  --max-num-batched-tokens 8192  --max-model-len 32768  --no-enable-prefix-caching  --async-scheduling  --compilation_config.pass_config.enable_fi_allreduce_fusion true --compilation_config.pass_config.enable_noop true  --compilation_config.custom_ops+=+rms_norm  --compilation_config.cudagraph_mode FULL_DECODE_ONLY  --compilation_config.splitting_ops []  --disable-log-requests -tp 2 `

`lm_eval --model local-completions --tasks gsm8k --model_args model=Qwen/Qwen3-Next-80B-A3B-Instruct,base_url=http://0.0.0.0:8000/v1/completions,max_retries=3,tokenized_requests=False,timeout=1200,max_gen_toks=2048,max_length=8192 --batch_size 2048 --trust_remote_code --limit 0.5`
## Test Result
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9530|±  |0.0082|
|     |       |strict-match    |     5|exact_match|↑  |0.9167|±  |0.0108|

```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by vadiklyutiy** [comment]
@jiahanc 
To double check, cutlass here means "vllm'a cutlass" or "FI's cutlass"?

**Comment by vadiklyutiy** [comment]
> > If for high concurrency trt-llm gem MoE is better maybe it's worth always enable it, not only for `VLLM_FLASHINFER_MOE_BACKEND=latency` ?
> 
> Yes, the trtllmgen moe is already the default FI MOE backend if FI MOE is enabled.

Oh, sorry, I just looked only into **Test Plan** section in the first msg and saw there `VLLM_FLASHINFER_MOE_BACKEND=latency`

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 1]
```diff
@@ -219,6 +238,67 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
             layer.w13_weight.data = shuffled_w13
             layer.w2_weight.data = shuffled_w2
 
+        if self.flashinfer_trtllm_moe_enabled:
+            from flashinfer.fused_moe.core import (
```
Comment (pavanimajety) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: Minor refactoring: can we move all the `self.flashinfer_trtllm_moe_enabled` post-weight-loading processing code block to flashinfer utilities?

Comment (pavanimajety) [reply vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: This might also benefit other precisions since this code/shuffling step is same for all trtllm moe flavors.

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 2]
```diff
@@ -58,6 +59,14 @@ def __init__(self, moe: FusedMoEConfig):
         else:
             self.rocm_aiter_fused_experts = None  # type: ignore
 
+        # Flashinfer TRTLLM MoE is only supported on Blackwell and later GPUS
+        self.flashinfer_trtllm_moe_enabled = (
+            has_flashinfer()
+            and envs.VLLM_USE_FLASHINFER_MOE_FP16
+            and envs.VLLM_FLASHINFER_MOE_BACKEND == "latency"
+            and current_platform.get_device_capability()[0] >= 10
```
Comment (vadiklyutiy) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: There is `has_device_capability()` that looks a little bit better

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 3]
```diff
@@ -58,6 +59,14 @@ def __init__(self, moe: FusedMoEConfig):
         else:
             self.rocm_aiter_fused_experts = None  # type: ignore
 
+        # Flashinfer TRTLLM MoE is only supported on Blackwell and later GPUS
+        self.flashinfer_trtllm_moe_enabled = (
+            has_flashinfer()
+            and envs.VLLM_USE_FLASHINFER_MOE_FP16
+            and envs.VLLM_FLASHINFER_MOE_BACKEND == "latency"
```
Comment (vadiklyutiy) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:69]: We discussed that _"Yes, the trtllmgen moe is already the default FI MOE backend if FI MOE is enabled."_
If so, why do we need this check of `VLLM_FLASHINFER_MOE_BACKEND`?

File: vllm/model_executor/layers/quantization/utils/flashinfer_utils.py [hunk 4]
```diff
@@ -284,7 +284,7 @@ def get_flashinfer_moe_backend() -> FlashinferMoeBackend:
     if flashinfer_moe_backend in backend_map:
         if (
             flashinfer_moe_backend == "latency"
-            and not current_platform.is_device_capability(100)
+            and not current_platform.has_device_capability(100)
```
Comment (vadiklyutiy) [inline vllm/model_executor/layers/quantization/utils/flashinfer_utils.py:293]: The same about _"Yes, the trtllmgen moe is already the default FI MOE backend if FI MOE is enabled."_

Shouldn't we remove `flashinfer_moe_backend` here?

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py (hunk 3 above)
Comment (jiahanc) [reply vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:69]: I think we should give user choice to choose between different MOE backends. Therefore we now set default backend be latency instead of totally remove the `VLLM_FLASHINFER_MOE_BACKEND`

File: vllm/model_executor/layers/quantization/utils/flashinfer_utils.py (hunk 4 above)
Comment (jiahanc) [reply vllm/model_executor/layers/quantization/utils/flashinfer_utils.py:293]: same with above

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py (hunk 3 above)
Comment (vadiklyutiy) [reply vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:69]: I agree that offering a choice is good behaviour.

However, the current naming is misleading. Suppose a user has a high-throughput use case; it is quite natural for them to set `VLLM_FLASHINFER_MOE_BACKEND == "throughput"`. By doing so, however, they end up using FlashInfer’s CUTLASS kernel, which is slower for high-throughput scenarios than the TRT-LLM generation backend. I am afraid this is not the desired behaviour.

Comment (jiahanc) [reply vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:69]: discussed offline, we will do a follow up PR about this

**Review by vadiklyutiy — APPROVED** [review]
Looks good for me

**Review by pavanimajety — APPROVED** [review]
Thanks for the PR and making the changes!

**Comment by vadiklyutiy** [comment]
@jiahanc 
I tried a bit different `lm_eval`
```
VLLM_USE_FLASHINFER_MOE_FP16=1 vllm serve Qwen/Qwen3-Next-80B-A3B-Instruct -tp 2 --enable-expert-parallel --async-scheduling --no-enable-prefix-caching
```
```
lm_eval --model local-completions --model_args model=Qwen/Qwen3-Next-80B-A3B-Instruct,base_url=http://localhost:8000/v1/completions -t gsm8k --num_fewshot 5 --batch_size 250
```

and got zero accuracy 
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.0174|±  |0.0036|
|     |       |strict-match    |     5|exact_match|↑  |0.0000|±  |0.0000|
```

**Comment by jiahanc** [comment]
> Seems the problem when `--enable-expert-parallel` added.

Thanks for finding the issue. EP was tested when related flashinfer PR just merged. After bisecting, the issue comes from a later PR in flashinfer https://github.com/flashinfer-ai/flashinfer/pull/2092, will convert current PR to draft and reopen when issue fixed.

**Comment by bnellnm** [comment]
> > EP accuracy was test at the time of development and accuracy was good so didnt test after rebase. Reverted back to commit before 1st rebase, seeing good accuracy, need to bisect the main branch commits
> 
> After bisect, after the refactor PR #28064, same code change cant work with expert parellel. @bnellnm may I ask do you have thoughts on what other code changes need after the refactor?



> > EP accuracy was test at the time of development and accuracy was good so didnt test after rebase. Reverted back to commit before 1st rebase, seeing good accuracy, need to bisect the main branch commits
> 
> After bisect, after the refactor PR #28064, same code change cant work with expert parellel. @bnellnm may I ask do you have thoughts on what other code changes need after the refactor?

Hi @jiahanc , I don't see anything obvious. #28064 mostly just moved classes out of layer.py and into individual files.  I would check to make sure that a message like "Swapping out UnquantizedFusedMoEMethod" does not appear in the DEBUG logs.  I don't think it should happen but that's the only thing I can think of that would cause something to break unexpectedly .

**Comment by bnellnm** [comment]
> --max-num-batched-tokens 8192 --max-model-len 32768

FYI, I think there are also some current CI problems with the Qwen3 model that may or may not be related to the accuracy problems you are seeing.

**Comment by vadiklyutiy** [comment]
> FYI, I think there are also some current CI problems with the Qwen3 model that may or may not be related to the accuracy problems you are seeing.

@bnellnm Qwen3 or Qwen3-next? Could you reference to the problems?
